add linked status indicator for scm inventory project updates

jakemcdermott opened the pull request:
##### SUMMARY
This adds a linked status indicator for scm inventory sync project updates. If the inventory update job is backed by an scm inventory source, we now show a second status icon and link for the related project update in a similar fashion to how we do for playbook job runs.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

##### ADDITIONAL INFORMATION
(updated)
![inventoryscm](https://user-images.githubusercontent.com/9753817/51689606-00908380-1fc5-11e9-8cda-8a7f37ac1fb8.gif)

jakemcdermott committed:
add linked status indicator for scm inventory project updates

Signed-off-by: Jake McDermott <yo@jakemcdermott.me>

jakemcdermott commented on awx/ui/client/features/output/output.strings.js:
```diff
@@ -22,6 +22,7 @@ function OutputStrings (BaseString) {
         EXPAND_OUTPUT: t.s('Expand Output'),
         EXTRA_VARS: t.s('Read-only view of extra variables added to the job template'),
         INVENTORY: t.s('View the Inventory'),
+        INVENTORY_SCM: t.s('View Project checkout results'),
```
same wording as the project update status icon for playbook jobs

dsesami reviewed (approved):
Ran a few iterations on this and tested it out for canceled jobs, failed sync behavior. Looks good, nice work.

AlanCoding commented:
Before this change, pretty scant on any project details:

![screen shot 2019-01-24 at 8 53 50 am](https://user-images.githubusercontent.com/1385596/51682745-282c1f80-1fb6-11e9-9879-28ef2588b1d7.png)

This looks great 👍 

I do realize that we don't provide a name for the source project / project update (they have the same name). If someone wanted that enhancement, then we would add it to the summary fields on the API side (which is trivial).

jakemcdermott commented:
@AlanCoding Makes sense - I've updated the serializer driving the details panel. Let me know what you think.

![inventoryscm](https://user-images.githubusercontent.com/9753817/51690039-e0ad8f80-1fc5-11e9-90a6-6790788690d3.gif)

I've also fixed a small bug I noticed while updating this where the project icon was showing for non-scm inventory update jobs. It should be good to go now.

@dsesami let me know if you'd like to take another pass at this once the new changes are cleared. To clarify the acceptance criteria a bit, the project status icon and linked name should only show for Inventory Update jobs that use the `scm` type inventory source.

AlanCoding commented:
Do the words "Ansible Playbook" in your gif link to the project?

jakemcdermott commented:
yep

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -2206,8 +2212,24 @@ def get_related(self, obj):
 
+    def get_summary_fields(self, obj):
+        summary_fields = super(UnifiedJobSerializer, self).get_summary_fields(obj)
+        source_project = getattrd(obj, 'source_project_update.unified_job_template', None)
```
seem like you would want `source_project = self.get_source_project(obj)` here

jakemcdermott replied:
this might be more to do with bad variable naming on my part - we want only the id to show up in the toplevel api data, so we return just the id with the `source_project` MethodField. 

When generating the summary fields we want the object data. I'm wondering if this variable would be better named as `source_project_obj` or something.

AlanCoding replied:
My preference would be that `get_source_project` returns the object and then if you need the id you do `get_source_project(obj).pk`. You can also do `getattrd(obj, 'source_project_update.unified_job_template_id', None)`, which sometimes has a performance benefit, but not in this case because you wind up fetching the object one way or the other.

jakemcdermott replied:
alright - `get_source_project` now returns the object

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -2176,11 +2176,17 @@ class Meta:
     custom_virtualenv = serializers.ReadOnlyField()
+    source_project = serializers.SerializerMethodField(
+        help_text=_('The project used for this job.'),
+    )
```
As a method field, this won't be filterable or anything like that, which someone might expect.

jakemcdermott replied:
How can we make it filterable?

AlanCoding replied:
You _can_ add the field and introduce a new migration for it. This also requires populating the field, which you would probably do at whatever point we already populate `source_project_update`. See:

https://github.com/ansible/awx/blob/51b51a9bf73de4a21c9aacb5c0c24f31bd44f04b/awx/main/tasks.py#L1703

You would add there `source_project=project_update.project`. I don't know _for sure_ right now that this is the only place that needs to be done.

jakemcdermott replied:
Hmm. I'm wondering if that's best left for another time - we didn't scope for much api testing on this and even though some might expect it to be filterable, its `filterable` field is correctly showing as `false` in the options data which means it is discoverable and correctly documented for the time being.

AlanCoding replied:
True, this might not be the best option for those reasons.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -2176,11 +2176,17 @@ class Meta:
+                  'source_project_update', 'custom_virtualenv', '-controller_node',)
+
+    def get_source_project(self, obj):
+        return getattrd(obj, 'source_project_update.unified_job_template.id', None)
```
This does incur additional queries, which is why I hesitated from doing it myself. You can add a new ForeignKey field, but that also automatically adds an index... which increases the memory footprint. My concern for these serializers is never the detail view, or a listing of project updates. My concern is always the `/api/v2/unified_jobs/` endpoint, which the UI hits aggressively. The additional queries will be hit for each SCM inventory update anywhere in the list as a user looks at the JOBS page.

jakemcdermott replied:
I see. How about we only do this for details views, then? I think there's a `is_detail` property we can leverage.

AlanCoding replied:
Yes, that is one workable option.

jakemcdermott replied:
alright - done. These are now only on the details serializer. I did some retesting and everything still looks good.

jakemcdermott commented:
@AlanCoding I think we should be good to go now - assuming unit tests pass I'll go ahead and run this through other tooling, e2e, etc.

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -2150,7 +2150,7 @@ class InventoryUpdateList(ListAPIView):
 
     model = InventoryUpdate
-    serializer_class = InventoryUpdateSerializer
+    serializer_class = InventoryUpdateDetailSerializer
```
There are still several occurrences of `InventoryUpdateSerializer` in this file, and I suspect you would want to change those to `InventoryUpdateDetailSerializer`. This is for the UI. Right on launch, it would probably anticipate getting the detail fields.

jakemcdermott replied:
👍 k. done.

AlanCoding reviewed (approved):
Since this is changing the top-level schema, I'd like to get at least 1 other API reviewer on it. I'm good with all this.

ryanpetrello reviewed (approved):
LGTM.  Neat!

mabashian reviewed (approved):
UI changes look good
